File: 148-sort-list/148-sort-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* merge(ListNode* l1, ListNode* l2){
        ListNode *dummy = new ListNode(-1);
        ListNode *curr = dummy;
        while(l1 && l2){
            if(l1->val < l2->val){
                curr->next = l1;
                l1 = l1->next;
            }
            else {
                curr->next = l2;
                l2 = l2->next;
            }
            curr = curr->next;
        }
        if(l1) curr->next = l1;
        if(l2) curr->next = l2;
        return dummy->next;
    }
    
    ListNode* sortList(ListNode* head) {
        if(!head || !head->next) return head;
        ListNode *slow=head, *fast=head->next;
        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        //diving the lists into 2 parts
        fast = slow->next;
        slow->next = NULL;
        //sort & merge
        head = sortList(head);
        fast = sortList(fast);
        return merge(head,fast);
    }
};

File: 148-sort-list/148-sort-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def merge(self,l1,l2):
        dummy = ListNode(-1)
        curr = dummy
        while(l1 and l2):
            if(l1.val<l2.val):
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            curr = curr.next
        if(l1):
            curr.next = l1
        if(l2):
            curr.next = l2
        return dummy.next
    
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if(not head or not head.next): return head
        slow,fast = head,head.next
        while(fast and fast.next):
            slow = slow.next
            fast = fast.next.next
        fast = slow.next
        slow.next = None
        head = self.sortList(head)
        fast = self.sortList(fast)
        return self.merge(head,fast)